pydicom/filereader.py

```diff
@@ -282,6 +294,12 @@ def data_element_generator(fp,
         if length != 0xFFFFFFFF:
             # don't defer loading of Specific Character Set value as it is
             # needed immediately to get the character encoding for other tags
+            if tag_set is not None and tag not in tag_set and tag != (
```

Comment:
Ok, I added tests for unspecified length.
@darcymason: I'm not sure, but AFAIK, undefined length happens only for sequence tags and pixel data - so I think the check for Specific Character Set in the undefined length branch (lines 378ff) can be removed. Is that correct?